### ISSUE TITLE
Rails 5.0.2 - Collection size error with counter_cache

### ISSUE THREAD

**Issue by silverdr** [rails/rails#28579]
### Steps to reproduce
```
class User < ApplicationRecord
	has_many :locks, class_name: 'UserLock', dependent: :destroy
end

class UserLock < ApplicationRecord
	belongs_to :user, counter_cache: :locks_count

	def locked?
		self.locks.size > 0
	end
end
```
Call `locked?()` when locks_count is set to NULL in the database.

### Expected behavior
Collection should be loaded and its length should be returned

### Actual behavior
`nil` is returned, which in turn causes `TypeError Exception: nil can't be coerced into Fixnum` being thrown.

In a debugger:
```
[106, 115] in /Users/silverdr/sources/silverdr/aria_main/app/models/user.rb
   106: 		self.activated_at.present?
   107: 	end
   108: 
   109: 	def locked?
   110: 		byebug
=> 111: 		self.locks.size > 0
   112: 	end
   113: 
   114: 	def deleted?
   115: 		self.deleted_at.present?
(byebug) self.locks.size
*** TypeError Exception: nil can't be coerced into Fixnum
```
Yet, when the locks are first loaded manually:
```
(byebug) self.locks
  UserLock Load (0.9ms)  SELECT "user_locks".* FROM "user_locks" WHERE "user_locks"."user_id" = $1  [["user_id", 1]]
#<ActiveRecord::Associations::CollectionProxy []>
(byebug) self.locks.size
0
(byebug) 
```
then correct result is returned.
### System configuration
**Rails version**: Rails 5.0.2

**Ruby version**: ruby 2.3.3p222 (2016-11-21 revision 56859) [x86_64-darwin15]

**Comment by rafaelfranca** [comment]
Could you create a reproduction script with this template https://github.com/rails/rails/blob/master/guides/bug_report_templates/active_record_master.rb?

**Comment by silverdr** [comment]
Yes, here you are.
```
begin
	require "bundler/inline"
rescue LoadError => e
	$stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
	raise e
end

gemfile(true) do
	source "https://rubygems.org"
	gem "rails", "5.0.2"
	gem 'sqlite3'
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
	create_table :user_locks, force: true do |t|
		t.references :user, foreign_key: true
	end

	create_table :users, force: true do |t|
		t.integer :locks_count
	end
end

class User < ActiveRecord::Base
	has_many :locks, class_name: 'UserLock', dependent: :destroy
end

class UserLock < ActiveRecord::Base
	belongs_to :user, counter_cache: :locks_count
end

class BugTest < Minitest::Test
	def test_association_stuff
		user = User.create!
		assert_equal 0, user.locks.size
	end
end
```

**Comment by rafaelfranca** [comment]
Thank you

**Comment by silverdr** [comment]
You are perfectly welcome. Since I am not familiar with current codebase, I don't feel like providing a pull-request at the moment but judging from the symptoms, it seems like for a non-loaded collection, the value from the database is returned directly, w/o validating (like e.g. checking it for being `nil`).

**Comment by maclover7** [comment]
Just checked and this is fixed on master. #26940 solved the issue.

**Comment by kamipo** [comment]
Fixed by #26940.
